File: 프로그래머스/unrated/155651. 호텔 대실/호텔 대실.py
import heapq


def solution(book_time):
    data = []
    answer = 1

    for b in book_time:
        s = int(b[0][:2])*60+int(b[0][3:])
        e = int(b[1][:2])*60+int(b[1][3:])
        data.append([s, e])

    data.sort()

    h = []
    for b in data:
        if not h:
            heapq.heappush(h, b[1])
            continue
        if h[0] <= b[0]:
            heapq.heappop(h)
        else:
            answer += 1
        heapq.heappush(h, b[1]+10)

    return answer

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.awt.Point;
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;

class Solution {
    public int solution(String[][] book_time) {
        List<Book> times = new ArrayList<>();
        for (String[] strings : book_time) {
            int start = getTime(strings[0]);
            int end = getTime(strings[1]);

            times.add(new Book(start, end));
        }

        times.sort((o1, o2) -> o1.start - o2.start);

        int answer = 1;
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (Book time : times) {
            if (pq.isEmpty()) {
                pq.offer(time.end + 10);
                continue;
            }

            if (pq.peek() <= time.start) {
                pq.poll();
            } else {
                answer++;
            }
            
            pq.offer(time.end + 10);
        }

        return answer;
    }

    public int getTime(String time) {
        String[] split = time.split(":");
        int h = Integer.parseInt(split[0]);
        int m = Integer.parseInt(split[1]);

        return h * 60 + m;
    }
}

class Book {
    int start;
    int end;

    public Book(int start, int end) {
        this.start = start;
        this.end = end;
    }
}